### PR TITLE
ARROW-15534: [C++] Add convenience function to substrait consumer to create plan instead of declaration

### PR DESCRIPTION
**Pull request by sanjibansg**
This PR adds a convenience function to Substrait consumer for the `DeserializePlan` method, which creates ExecPlan for single root relations.

### CODE REVIEW

File: cpp/src/arrow/engine/substrait/serde.cc [hunk 1]
```diff
@@ -86,6 +86,20 @@ Result<std::vector<compute::Declaration>> DeserializePlan(
   return sink_decls;
 }
 
+Result<compute::ExecPlan> DeserializePlan(const Buffer& buf,
+                                          const ConsumerFactory& consumer_factory,
+                                          ExtensionSet* ext_set_out) {
+  ARROW_ASSIGN_OR_RAISE(auto declarations,
+                        DeserializePlans(buf, consumer_factory, ext_set_out));
+  if (declarations.size() > 1) {
+    return Status::Invalid("Substrait plan cannot have multiple root relations");
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/serde.cc:95]:
```suggestion
    return Status::Invalid("DeserializePlan does not support root relations");
```
Technically a Substrait plan can have multiple root relations.  We just don't support those plans in this method.

File: cpp/src/arrow/engine/substrait/serde.h [hunk 2]
```diff
@@ -48,9 +48,12 @@ using ConsumerFactory = std::function<std::shared_ptr<compute::SinkNodeConsumer>
 /// returned here.
 /// \return a vector of ExecNode declarations, one for each toplevel relation in the
 /// Substrait Plan
-ARROW_ENGINE_EXPORT Result<std::vector<compute::Declaration>> DeserializePlan(
-    const Buffer& buf, const ConsumerFactory& consumer_factory,
-    ExtensionSet* ext_set = NULLPTR);
+ARROW_ENGINE_EXPORT Result<std::vector<compute::Declaration>> DeserializePlans(
+    const Buffer& buf, const ConsumerFactory& consumer_factory, ExtensionSet* ext_set);
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/serde.h:52]:
```suggestion
    const Buffer& buf, const ConsumerFactory& consumer_factory, ExtensionSet* ext_set = NULLPTR);
```
I think we still need the default value here.

File: cpp/src/arrow/engine/substrait/serde.cc [hunk 3]
```diff
@@ -86,6 +86,20 @@ Result<std::vector<compute::Declaration>> DeserializePlan(
   return sink_decls;
 }
 
+Result<compute::ExecPlan> DeserializePlan(const Buffer& buf,
+                                          const ConsumerFactory& consumer_factory,
+                                          ExtensionSet* ext_set_out) {
```
Comment (westonpace) [inline cpp/src/arrow/engine/substrait/serde.cc:91]: I do like the name `ext_set_out` more than `ext_set`.  Could you update the name in the header as well so they match?

**Review by westonpace — CHANGES_REQUESTED** [review]
Just a few suggestions but looks good.

File: cpp/src/arrow/engine/substrait/serde.cc (hunk 1 above)
Comment (sanjibansg) [reply cpp/src/arrow/engine/substrait/serde.cc]: Yes, right, made the change, thanks!

File: cpp/src/arrow/engine/substrait/serde.h (hunk 2 above)
Comment (sanjibansg) [reply cpp/src/arrow/engine/substrait/serde.h]: Made the change.

File: cpp/src/arrow/engine/substrait/serde.cc (hunk 3 above)
Comment (sanjibansg) [reply cpp/src/arrow/engine/substrait/serde.cc:91]: Renamed ext_set to ext_set_out.

**Review by westonpace — APPROVED** [review]
One small change and we can merge on green.

File: cpp/src/arrow/engine/substrait/serde.cc [hunk 4]
```diff
@@ -86,6 +86,20 @@ Result<std::vector<compute::Declaration>> DeserializePlan(
...
+  ARROW_ASSIGN_OR_RAISE(auto declarations,
+                        DeserializePlans(buf, consumer_factory, ext_set_out));
+  if (declarations.size() > 1) {
+    return Status::Invalid("DeserializePlan does not support root relations");
```
[the remark this answers is not in the document]
Comment (sanjibansg) [reply cpp/src/arrow/engine/substrait/serde.cc]: Corrected that, thanks!

**Pull request closed without merging**